### PR TITLE
PP-275 HTTP connection leak in Connector

### PR DESCRIPTION
**Pull request by ckalista**
# WHAT
- New GatewayClient.Response pre-fetches the entity of the response it wraps and deletegates all other operation.
- Enforce the response close in a finally before returning to protect against memory leaks.
# WHO

Any dev

### CODE REVIEW

File: src/main/java/uk/gov/pay/connector/service/GatewayClient.java [hunk 1]
```diff
@@ -77,11 +84,15 @@ public static GatewayClient createGatewayClient(Client client, String gatewayUrl
         } catch (Exception e) {
             logger.error(format("Exception for gateway url=%s", gatewayUrl), e);
             return left(baseError(e.getMessage()));
+        } finally {
```
Comment (imranbohoran) [inline src/main/java/uk/gov/pay/connector/service/GatewayClient.java]: 👍

File: src/main/java/uk/gov/pay/connector/service/GatewayClient.java [hunk 2]
```diff
@@ -91,4 +102,98 @@ public static GatewayClient createGatewayClient(Client client, String gatewayUrl
             return left(malformedResponseReceivedFromGateway("Invalid Response Received From Gateway"));
         }
     }
+
+    static public class Response {
```
Comment (imranbohoran) [inline src/main/java/uk/gov/pay/connector/service/GatewayClient.java]: I still can't think of a good reason to keep these unused methods exposed form the new Response object. We only ever seem to be using getEntity which is what is used to unmarshall into the object, and the getStatus method. We can add them if needed (although I don't see a need arising for them to be added)

File: src/main/java/uk/gov/pay/connector/service/GatewayClient.java [hunk 3]
```diff
@@ -91,4 +102,98 @@ public static GatewayClient createGatewayClient(Client client, String gatewayUrl
...
+        private final javax.ws.rs.core.Response delegate;
+        private final String entity;
+
+        protected Response(final javax.ws.rs.core.Response delegate) {
+            this.delegate = delegate;
+            this.entity = delegate.readEntity(String.class);
+        }
+
+        public int getStatus() {
+            return delegate.getStatus();
```
Comment (imranbohoran) [inline src/main/java/uk/gov/pay/connector/service/GatewayClient.java]: Shouldn't this be initialised at the constructor, as the delegate (whichi is the response) might be closed by the time this method is invoked?

Comment (ckalista) [inline src/main/java/uk/gov/pay/connector/service/GatewayClient.java]: My understanding is that only the entity methods where the entity is backed by an input stream are affected by the close.

File: src/main/java/uk/gov/pay/connector/service/GatewayClient.java (hunk 2 above)
Comment (ckalista) [inline src/main/java/uk/gov/pay/connector/service/GatewayClient.java]: If we didn't need some of those methods we could have probably gotten away with returning a String. However we do seem to need more control. For instance here: 

```
    private InquiryResponse inquire(String transactionId, GatewayAccountEntity gatewayAccount) {
        return reduce(
                client
                        .postXMLRequestFor(gatewayAccount, buildOrderInquiryFor(gatewayAccount, transactionId))
                        .bimap(
                                InquiryResponse::inquiryFailureResponse,
                                (response) -> response.getStatus() == OK.getStatusCode() ?
                                        mapToInquiryResponse(response) :
                                        errorInquiryResponse(logger, response)
                        )
        );
    }
```

Comment (imranbohoran) [inline src/main/java/uk/gov/pay/connector/service/GatewayClient.java]: Yes, but isn't that just getStatus(). So we only expose getEntity() and getStatus() and if we initialise the entity and status in the constructor there's no need to do any delegation.

Comment (ckalista) [inline src/main/java/uk/gov/pay/connector/service/GatewayClient.java]: The goal was not to close the API but as discussed I am not fussed so happy to change that.

**Comment by ckalista** [comment]
rebuild

**Comment by imranbohoran** [comment]
👍 Looks good

**Pull request merged**